File: Programmers/codingTest_Kit/Stack_Queue/42586.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

vector<int> solution(vector<int> progresses, vector<int> speeds) {
    vector<int> answer;
    queue<int> Q;
    for(int i = 0; i < progresses.size(); i++){
        int leftRate = 100 - progresses[i];
        int date = leftRate / speeds[i];
        if(leftRate % speeds[i] != 0) date += 1;
        Q.push(date);        
    }
    int start = Q.front();
    int count = 0;
    while(!Q.empty()){
        if(Q.front() <= start){
            Q.pop();
            count++;
        }
        else{
            if(count > 0){
                start = Q.front();
                answer.push_back(count);
                count = 0;
            }
        }
    }
    answer.push_back(count);
    return answer;
}

File: Programmers/Practice/LV2/42586.java
package LV2;

import java.util.*;

class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        Queue<Integer> que = new ArrayDeque<>();
        List<Integer> answer = new ArrayList<>();

        for (int i = 0; i < progresses.length; i++) {
            int leftRate = 100 - progresses[i];
            int date = leftRate / speeds[i];
            if (leftRate % speeds[i] != 0) {
                date += 1;
            }

            que.add(date);
        }

        int start = que.peek();
        int count = 0;

        while (!que.isEmpty()) {
            if (que.peek() <= start) {
                que.remove();
                count++;
            } else {
                if (count > 0) {
                    start = que.peek();
                    answer.add(count);
                    count = 0;
                }
            }
        }

        answer.add(count);

        return answer.stream().mapToInt(Integer::intValue).toArray();
    }

}